### ISSUE TITLE
Order of packages listed in environment.yml yields different results depending on the order

### ISSUE THREAD

**Issue by nicoddemus** [conda/conda#4957]
Hi,

Consider this `environment.yml` file:

```yml
name: test-new-conda
dependencies:  
  - pytest>2
  - pytest==3.0.5
  - python==2.7.12
```

When executing `conda env update --prune`, the pytest version that gets installed is `2.9.2+1` instead of the expected `3.0.5`:

```
λ conda list -n test-new-conda
# packages in environment at E:\Miniconda\envs\test-new-conda:
#
bzip2                     1.0.6           vc10_esss_dist_1  [vc10]  esss
colorama                  0.3.7                    py27_0    mirror
db                        5.2.0           vc10_esss_dist_1  [vc10]  esss
openssl                   1.0.0e          vc10_esss_dist_1  [vc10]  esss
pip                       9.0.1                    py27_1    mirror
py                        1.4.33                   py27_0    mirror
pytest                    2.9.2+1                  py27_0    esss
python                    2.7.12                        2    esss
setuptools                12.2            py27_esss_dist_1    esss
sqlite                    3070900         vc10_esss_dist_1  [vc10]  esss
vs2010_runtime            10.00.40219.1                 2    mirror
wheel                     0.29.0                   py27_0    mirror
zlib                      1.2.5           vc10_esss_dist_2  [vc10]  esss
```

However, if I put `pytest==3.0.5` first:

```yml
name: test-new-conda
dependencies:  
  - pytest==3.0.5
  - pytest>2  
  - python==2.7.12
```

Then the expected `pytest` version is installed:

```
λ conda list -n test-new-conda
# packages in environment at E:\Miniconda\envs\test-new-conda:
#
bzip2                     1.0.6           vc10_esss_dist_1  [vc10]  esss
colorama                  0.3.7                    py27_0    mirror
db                        5.2.0           vc10_esss_dist_1  [vc10]  esss
openssl                   1.0.0e          vc10_esss_dist_1  [vc10]  esss
pip                       9.0.1                    py27_1    mirror
py                        1.4.33                   py27_0    mirror
pytest                    3.0.5                    py27_0    mirror
python                    2.7.12                        2    esss
setuptools                12.2            py27_esss_dist_1    esss
sqlite                    3070900         vc10_esss_dist_1  [vc10]  esss
vs2010_runtime            10.00.40219.1                 2    mirror
wheel                     0.29.0                   py27_0    mirror
zlib                      1.2.5           vc10_esss_dist_2  [vc10]  esss
```

Version `2.9.2+1` is a package released in our internal channel to test a new pytest release.

`conda-4.2.16` installs `pytest-3.0.5` as expected using both versions of the `environment.yml` file.

Here's my `conda info`:

```
λ conda info
Current conda install:

               platform : win-64
          conda version : 4.3.14
       conda is private : False
      conda-env version : 4.3.14
    conda-build version : 2.1.5
         python version : 2.7.9.final.0
       requests version : 2.13.0
       root environment : E:\Miniconda  (writable)
    default environment : E:\Miniconda
       envs directories : E:\Miniconda\envs
                          C:\Users\bruno\AppData\Local\conda\conda\envs
                          C:\Users\bruno\.conda\envs
          package cache : E:\Miniconda\pkgs
                          C:\Users\bruno\AppData\Local\conda\conda\pkgs
           channel URLs : https://conda-channel/esss/win-64
                          https://conda-channel/esss/noarch
                          https://conda-channel/mirror/win-64
                          https://conda-channel/mirror/noarch
                          https://conda-channel/mirror-conda-forge/win-64
                          https://conda-channel/mirror-conda-forge/noarch
            config file : E:\Miniconda\.condarc
           offline mode : False
             user-agent : conda/4.3.14 requests/2.13.0 CPython/2.7.9 Windows/8 Windows/6.2.9200

```

(Please note that the channels above can't be accessed from outside our company).

Changing `conda` versions, I can pinpoint that the behavior changed from version `4.3.4` (which installs the correct pytest) to `4.3.5` (installs the incorrect one).

It seems this was a bug introduced by accident.

cc @gqmelo @tadeu @prusse-martin

**Comment by goanpeca** [comment]
@nicoddemus why would one add a package twice on the list with different specification? This should raise an error

`DuplicatePackageError` or something, no?

**Comment by nicoddemus** [comment]
@goanpeca thanks for the quick response!

We are not actually adding the package twice, that was meant to simplify the example; we actually are seeing this error because one of our `environment.yml` files included several pytest plugins before pytest itself. 

For example:

```yaml
name: test-new-conda
dependencies: 
  - pytest-cov
  - pytest==3.0.5  
  - python==2.7.12
```

This file installs `pytest-3.0.5` on `4.2.16` and `pytest-2.9.2+1` on `4.3.14`.

**Comment by goanpeca** [comment]
Ahh ok, thats a much better example :-), will look into it

**Comment by nicoddemus** [comment]
Thanks!

**Comment by nicoddemus** [comment]
@goanpeca It is easy to reproduce the issue using standard packages and local packages:

```yaml
- file:///C:/tmp/channel1/win-64
  - pytest-2.9.0-py27_0.tar.bz2
  - pytest-cov-2.3.1-py27_0.tar.bz2
- file:///C:/tmp/channel2/win-64
  - pytest-3.0.5-py27_0.tar.bz2
- defaults
```

**Comment by goanpeca** [comment]
@mcg1969, @kalefranz ?

**Comment by kalefranz** [comment]
Fixed recently in 4.3.17.

**Comment by nicoddemus** [comment]
Thanks for the ping!
